### ISSUE TITLE
Git task cannot handle special versions such as 1.0.0beta1

### ISSUE THREAD

**Issue by jennings** [jennings/SemVerHarvester#2]
Rule 3 of the semantic versioning spec says:

> A special version number MAY be denoted by appending an arbitrary string immediately following the patch version. The string MUST be comprised of only alphanumerics plus dash [0-9A-Za-z-] and MUST begin with an alpha character [A-Za-z]. Special versions satisfy but have a lower precedence than the associated normal version.

The task currently does not handle this at all.

**Comment by marfarma** [comment]
Of the issues listed, this is the one that interests me most. Did you have any thoughts on how to approach it?

**Comment by jennings** [comment]
For a tag like "v1.2.3beta1", I think it should still do `AssemblyVersion("1.2.3.x")` as though the "beta1" weren't there. I think it would only affect `AssemblyInformationalVersion`, since that one accepts a string.

**Comment by Grynn** [comment]
That sounds like a good idea; may also be useful to capture other state - like 
'dirty' => source had untracked/modified files at time of build
'commit' => id of last commit
'branch' => branch

and stick all of them in a `AssemblyConfiguration` attribute.

**Comment by marfarma** [comment]
@Grynn - we already have the property ModifiedString that takes care of the 'dirty' thing.  I'd like to have the 'commit' and 'branch' properties returned too.  However, I think that discussion should be a separate ticket.  There are valid arguments against expanding SemVerHarvester away from its core function.   

I like that the current implementation doesn't generate the AssemblyConfiguration attribute.  I'm not sure, but are you suggesting that is should?  I'd prefer we update our documentation to offer suggestion, one of which would cover assigning an OptVersionPart property to the AssemblyInformationalVersion.

@jennings - I agree we should still bring back the triple and the ModifiedString as they are now.  For any optional elements I'm thinking we could support a regular expression parameter, defaulted to capture the balance of the tag name.  However, if provided, a custom match pattern might support pulling back 'beta' and '1' as separate values.  I don't know what's possible in project config files.  I suspect arrays are not. If so, we could create  OptVersionPart1 through OptVersionPart5 - and validate any submitted match pattern has 5 or fewer captures.

Probably best if we start by implementing the default case.  Afterwards, we could look at supporting custom match patterns, and additional OptVersionParts.

**Comment by jennings** [comment]
Should we be touching AssemblyConfiguration? My understanding is that it's used to indicate which _type_ of assembly was built (with debugging symbols, Release build, etc.). I'm not sure it's appropriate to put version information in there.

Regarding special versions: SemVer says you can put a prerelease version or a build number after the patch version. The former is preceeded with a dash, the latter is preceeded by a plus. So a build is of the format:

```
SemVer 2.0.0-rc.1:
version = "v", major, [".", minor, [".", patch]], ["-", prerelease], ["+", build]

SemVer 1.0.0:
version = "v", major, [".", minor, [".", patch]], [["-"], prerelease]
```

Where "major", "minor", and "patch" are integers, and "prerelease" and "build" are arbitrary strings excluding dashes and pluses (in SemVer 1.0, any character is allowed).

It might be easiest to just treat everything after the patch number as an arbitrary string and include it in AssemblyInformationalVersion.

**Comment by marfarma** [comment]
> It might be easiest to just treat everything after the patch number as an arbitrary string and include it in AssemblyInformationalVersion.

I don't think we should put anything in AssemblyInformationalVersion -- I think that we should return the arbitrary string as an additional property, and let the user add the string to their AssemblyInformationalVersion, or not, as they choose, as we do now with major, minor and patch.

Other than that - I agree.

**Comment by jennings** [comment]
Whoops, haha, you're right. It's been so long since I worked on this that I forgot what it did.

You're right, I don't think this MSBuild task should do anything except return the version information. Modifying assembly attributes should be for the user to decide ([example here](https://github.com/jennings/OATH.Net/blob/master/OATH.Net/OATH.Net.csproj#l60-72)).
